### ISSUE TITLE
`not()` with NULL literals does not work: can't be evaluated because the expression's type is Utf8, not boolean

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#1191]
Reproducer:

```sql
> select not null;
Internal("NOT 'Literal { value: Utf8(NULL) }' can't be evaluated because the expression's type is Utf8, not boolean")

> select not(null);
Internal("NOT 'Literal { value: Utf8(NULL) }' can't be evaluated because the expression's type is Utf8, not boolean")

```

The correct output is `null`

I think the solution will likely be to teach `Expr::Not` about type coercion (as is done in `BinaryExpr` and function calls) once coercion knows about the null type (and that it can be cast to any other type). See https://github.com/apache/arrow-datafusion/issues/1184 for more details

**Comment by WinkerDu** [comment]
@alamb plz assign this issue to me, thank you.

**Comment by WinkerDu** [comment]
A raw fix here: #2144
